### ISSUE TITLE
[Bug]: Some protocol violations or bugs in VerneMQ 

### ISSUE THREAD

**Issue by songxpu** [vernemq/vernemq#2283]
### Environment

- VerneMQ Version: 2.0.0-rc1-5-gf0e6dc15
- OS: Ubuntu 20.04
- Erlang/OTP version (if building from source):
- Cluster size/standalone:

### Current Behavior

VerneMQ receives such a request and returns a successful response message or processes and saves it normally

### Expected behaviour

In these scenarios, VerneMQ receives such a request with the expectation that it should abort the connection or reject the message

### Configuration, logs, error output, etc.

```markdown
plz set `allow_anonymous = on`, and most settings are default
```

### Code of Conduct

- [x] I agree to follow the VerneMQ's Code of Conduct

**Comment by mths1** [comment]
> According to the specification of [MQTTv5.0](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html):
> 
> ```
> [MQTT-3.3.4-6] 
> A PUBLISH packet sent from a Client to a Server MUST NOT contain a Subscription Identifier.
> ```
> 
> However, if we send such a packet including Connect and Publish:
> 
> ```
> 106100044d5154540540b3b037119afb60e317001901215c5326000f766e72366d4541644d78553c44327800173049574d36324268715a6179524b5a62536749534a31360013317675535434755733374e64397846585a38570008676235716d583636
> 3554000841684b3146454962f71a16010109000c79a506aff5eef39ed5210cd60bba849b523155596942337761334c376e765936573739413862666a46414e4e3172647544345773415778724a6667386d3258653363
> ```

@ioolkos : This one is interesting. I can reproduce, but it seems that the publish is never executed. I can see valid MQTT in Wireshark...

**Comment by ioolkos** [comment]
@mths1 yeah, the spec is relatively clear that it's a protocol error. So, I guess we should close the connection.
(Not sure I understand "the publish is never executed" correctly).

---
👉 Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
👉 Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by mths1** [comment]
@ioolkos : It means a subscribe on # won't show anything and I set a debug message on the "publish" frame which never seems to be executed.

**Comment by mths1** [comment]
@ioolkos : Regarding the  Packet Identifier > 0 problem, would be easy to fix but another check on each publish. I am honestly not sure why the spec does not allow 0 :-) (other than philosophical discussions about if 0 is a number or not)

**Comment by mths1** [comment]
@songxpu : Yeah, I should have been more detailed here. When I do two requests it works as expected (or not expected :-) ). I was not sure what the correct behaviour is in the "all at once" case.

**Comment by ioolkos** [comment]
> @songxpu : Yeah, I should have been more detailed here. When I do two requests it works as expected (or not expected :-) ). I was not sure what the correct behaviour is in the "all at once" case.

I think it's related to the fact that the tests here don't use an MQTT client state machine. Clients wait to get a CONNACK from the server before they send messages.

Although.... technically clients are allowed to send PUBLISH's to the server right away, that is: before getting a CONNACK (at their own risk because the messages will be lost if the client gets rejected). I need to verify whether Verne accepts PUBLISHs at all before sending out a CONNACK and have the MQTT session fully in connected state. Note that the problem with that is that the server has to "hold back" processing the PUBLISHs until it has confirmed he connection; otherwise it would have processed messages from an unauthorized client.

---
👉 Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
👉 Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by mths1** [comment]
@songxpu : Thanks for the report. All issues should be addressed by the linked pull requests. Feel free to test them if you want. May I ask what is the background of you performing all those tests? (feel free to do more, though :-) )

**Comment by ioolkos** [comment]
@songxpu to fully remove *any* state on a VerneMQ node (for example between test runs), you can `mv` away or delete everything that's under the VerneMQ `data` directory. VerneMQ will then recreate all the needed directories upon boot; all state will be fresh (no subscriptions, messages, cluster state etc.)

---
👉 Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
👉 Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by ioolkos** [comment]
@songxpu all the fixes are now merged. Do you want to keep this thread open, or shall we close it for now?
Thanks to @mths1 for the contributions!

---
👉 Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
👉 Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by ioolkos** [comment]
@songxpu thanks for the additional info.

The real question is what to do when we'd find a wildcard character in a response topic. Still deliver the message? There's also no appropriate Response code for this in the Pubrec response codes.
But let's see what we can do.

---
👉 Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
👉 Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.
